src/v037/codec.rs

```diff
@@ -0,0 +1,111 @@
+use std::marker::PhantomData;
+
+use tokio_util::codec::{Decoder, Encoder};
+
+use bytes::{BufMut, BytesMut};
+
+pub struct Decode<M> {
+    state: DecodeState,
+    _marker: PhantomData<M>,
+}
+
+impl<M> Default for Decode<M> {
+    fn default() -> Self {
+        Self {
+            state: DecodeState::Head,
+            _marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DecodeState {
+    Head,
+    Body { len: usize },
+}
+
+impl<M: prost::Message + Default> Decoder for Decode<M> {
+    type Item = M;
+    type Error = crate::BoxError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        match self.state {
+            DecodeState::Head => {
+                tracing::trace!(?src, "decoding head");
+
+                // we don't use decode_varint directly, because it advances the
+                // buffer regardless of success, but Decoder assumes that when
+                // the buffer advances we've consumed the data. this is sort of
+                // a sad hack, but it works.
+                // TODO(erwan): fix this
```

Comment:
I don't know that this needs to be fixed, as such -- the code is ugly, but it works, and is actually pretty efficient, since we're just making a clone of the "view" into the buffer, letting it be advanced by the decode method, and then discarding it.